src/analyzer/protocol/websocket/websocket.spicy

```diff
@@ -0,0 +1,121 @@
+# Copyright (c) 2024 by the Zeek Project. See LICENSE for details.
+
+module WebSocket;
+
+import spicy;
+import zeek;
+
+const OPCODE_CONTINUATION = 0x00;
+const OPCODE_TEXT         = 0x01;
+const OPCODE_BINARY       = 0x02;
+const OPCODE_CLOSE        = 0x08;
+const OPCODE_PING         = 0x09;
+const OPCODE_PONG         = 0x0a;
+
+# Are bytes copied like a string in C++, or is that ref counted and cheap
+# like it would be in Python?
+function unmask(inout f: Frame, chunk: bytes): bytes {
+  if ( ! f.mask )
+    return chunk;
+
+  local unmasked_chunk: bytes;
+  local i = 0;
+  # This is very very expensive :-( Creates a new shared_ptr for
+  # each byte we iterate over it seems. Visible in the first
+  # flamegraph I looked at.
+  for ( c in chunk ) {
+    unmasked_chunk += c ^ f.masking_key[f.masking_key_idx % 4];
+    i++;
+    f.masking_key_idx++;
+  }
+
+  return unmasked_chunk;
+}
+
+type Frame = unit(m: Message) {
+  var payload_len: uint64;
+  var masking_key_idx: uint64;
+  var data: bytes;
+  var close_data: bytes;
+  var effective_opcode: uint8;
+
+  : bitfield(16) {
+    fin: 0 &convert=cast<bool>($$);
+    rsv: 1..3;
+    opcode: 4..7 &convert=cast<uint8>($$);
+    mask: 8 &convert=cast<bool>($$);
+    payload_len1: 9..15;
+  } &bit-order=spicy::BitOrder::MSB0;
+
+  : void &requires=(m.opcode != 0 && self.opcode == 0 || m.opcode == 0 && self.opcode != 0) {
+    self.effective_opcode = m.opcode != 0 ? m.opcode : self.opcode;
+  }
+
+  payload_len2: uint16 if (self.payload_len1 == 126);
+  payload_len8: uint64 if (self.payload_len1 == 127);
+
+  : void {
+    self.payload_len = self.payload_len1;
+    if ( self?.payload_len2 )
+      self.payload_len = self.payload_len2;
+    else if ( self?.payload_len8 )
+      self.payload_len = self.payload_len8;
+  }
+
+  masking_key: uint8[] &size=4 if (self.mask);
```

Comment:
This being an `uint8[]` allows `masking_key[x]` indexing, while a `bytes` object would require `*masking_key.at(i)` which took roughly 20% more runtime when I tested it.